feat: migration("re-indexing"), backfilling and diasgnostics tooling for the `ChainIndexer` (#12450)
filecoin-project/lotus: chain/index/README.MD

Reviewer: There is inlined documentation like API documentation and command line tool docs.  These are going to get stale.  I think it's better to link to their source of truth.  (I will point this out where I see it.)

Reviewer: I expected to see a rollback section but didn't see one.

Reviewer: Ok, sounds good.  I assume you'll resolve this conversation once it's moved out?

Reviewer: Tracked in https://github.com/filecoin-project/lotus/pull/12600.  Resolving.

# ChainIndexer Documentation for RPC Providers

## Introduction

> We're shipping a new Indexer implementation in Lotus (`ChainIndexer`) to index Filecoin chain state such as tipsets, messages, events and ETH transactions for reliable and faster RPC responses. The `ChainIndexer` replaces the existing `MsgIndex`, `EthTxIndex` and `EventIndex` implementations in Lotus which suffer from a multitude of known problems documented [here](https://github.com/filecoin-project/lotus/issues/12293). 

Author: @rjan90 Can we do it closer to the release ? Can keep this comment unresolved till then.

> We're shipping a new Indexer implementation in Lotus (`ChainIndexer`) to index Filecoin chain state such as tipsets, messages, events and ETH transactions for reliable and faster RPC responses. The `ChainIndexer` replaces the existing `MsgIndex`, `EthTxIndex` and `EventIndex` implementations in Lotus which suffer from a multitude of known problems documented [here](https://github.com/filecoin-project/lotus/issues/12293). 

Author: Agreed. Moved this detail to the Changelog and linked this document there.

> We're shipping a new Indexer implementation in Lotus (`ChainIndexer`) to index Filecoin chain state such as tipsets, messages, events and ETH transactions for reliable and faster RPC responses. The `ChainIndexer` replaces the existing `MsgIndex`, `EthTxIndex` and `EventIndex` implementations in Lotus which suffer from a multitude of known problems documented [here](https://github.com/filecoin-project/lotus/issues/12293). 

Reviewer: No longer relevant as the docs have been moved other places.  Resolving.

## ChainIndexer Config

The `ChainIndexer` must be enabled on an RPC node as it is disabled by default. Here is the mandatory config to use for all RPC providers to enable the `ChainIndexer` and for ensuring the `EthRPC` and `ActorEventsAPI` are enabled:

```toml
[ChainIndexer]
# This is set to false by default which disables the ChainIndexer. 
# Please ensure that you set it to true before starting your node.
  EnableIndexer = true 

[Fevm]
# This is set to false by default which disables the ETH RPC API. 
# Please ensure that you set it to true before starting your node.
  EnableEthRPC = true

[Events]
# This is set to false by default which disables the Actor Events API. 
# Please ensure that you set it to true before starting your node.
  EnableActorEventsAPI = true
```

### Garbage Collection

The `ChainIndexer` includes a garbage collection (GC) mechanism to manage the amount of historical data retained. By default, GC is disabled to preserve all indexed data.

#### Configuration

To configure GC, use the `GCRetentionEpochs` parameter in the `ChainIndexer` section of your config.

> The ChainIndexer periodically runs GC if `GCRetentionEpochs` is > 0 and removes indexed data for epochs older than `(current_head_height - GCRetentionEpochs)`.

Author: @BigLep No this isn't configurable right now. It's very easy to do so but I really want to ship with a minimal config here and then iterate based on user feedback rather than shipping a bloated config where most options are never used by users but do add to the complexity of the config.

It's a trivial change to make if we do get a strong ask for this.

> The ChainIndexer periodically runs GC if `GCRetentionEpochs` is > 0 and removes indexed data for epochs older than `(current_head_height - GCRetentionEpochs)`.

Reviewer: Understood that this isn't something we want to expose as configurable unless there is ask for it.  
And it sounds like we don't know if users care about how frequently it runs.
I think the path to take it to point to the source of truth.  In the updated revision I added this as a suggestion.  I'll resolve this conversation.

```toml
[ChainIndexer]
  GCRetentionEpochs = X  # Replace X with your desired value
```

- Setting `GCRetentionEpochs` to 0 (**this is the default**) completely disables GC.

> - Any non-zero value enables GC and determines the number of epochs of historical data to retain.

Author: @BigLep We don't have automated tests yet but this will be covered when we write them yes.

#### Recommendations

1. **Archival Nodes**: **Keep GC disabled** (`GCRetentionEpochs` = 0) to retain all indexed data.

2. **Non-Archival Nodes**:  Set `GCRetentionEpochs` to match the amount of chain state your node retains 

> (*for example:* if your node is configured to retain 2 days of Filecoin chain state with the Splitstore, set `GCRetentionEpochs` to (number of Filecoin epochs in a day *2) = 5760).

Author: @BigLep The Splitstore config is hard to reason about. But RPC providers have a good understanding of how much history is maintained on their nodes because of experience running it. So, best to keep this as it is.

> (*for example:* if your node is configured to retain 2 days of Filecoin chain state with the Splitstore, set `GCRetentionEpochs` to (number of Filecoin epochs in a day *2) = 5760).

Reviewer: Ack, got it.  Makes sense.  

Improvement for slight readability improvement:

```
(*Example:* if your node is configured to retain 2 days of Filecoin chain state with the Splitstore, set `GCRetentionEpochs` to `retentionDays * epochsPerDay = 2 * 2880 = 5760`.)
````

> (*for example:* if your node is configured to retain 2 days of Filecoin chain state with the Splitstore, set `GCRetentionEpochs` to (number of Filecoin epochs in a day *2) = 5760).

Reviewer: I'll suggest this in the updated revision and resolve here.

> ---

Reviewer: We're under a "ChainIndexer Configuration" section.  How about making this it's own subheading?

```suggestion
### Removed Options
```

Reviewer: Removed configuration options feels like content that should be in a changelog.

**Note: The following config options no longer exist in Lotus and have been removed in favor of the `ChainIndexer` config options explained above:**

```toml
[Fevm]
EthTxHashMappingLifetimeDays = 0

[Events]
DisableHistoricFilterAPI = false
DatabasePath = ""

> ```

Reviewer 2: Maybe change it the Note to call out that these legacy configs can be removed when upgrading to the new version:

**Note: The following config options no longer exist in Lotus and have been removed in favor of the `ChainIndexer` config options explained above. They can be removed when upgrading to the Lotus vX.XX.X version"

Author: Done. Let's fill in the `vX.XX.X` part when we get closer to the release.

These options are now deprecated and will not have any effect if used in your configuration file. Please use the `ChainIndexer` config options as described above.

> ---

Reviewer: ```suggestion
```

I think can be removed if we have a "removed" configuration option.

## Migration Guide

> Migrating to the new `ChainIndexer` involves several steps to ensure a smooth transition. Here's a guide to help you through the process:

Reviewer 2: Time estimates for the migration for Snapshot synced and archival would also be good to add at the top level here.

Author: @rjan90 Agreed but I have yet to test on a mainnet snapshot/Archival node. Let's add this once I've gotten around to doing it and have the required numbers. Keeping this comment open.

Reviewer: This is being handled in https://github.com/filecoin-project/lotus/pull/12600

> Migrating to the new `ChainIndexer` involves several steps to ensure a smooth transition. Here's a guide to help you through the process:

Reviewer: ```suggestion
Migrating to the new `ChainIndexer` involves several steps to ensure a smooth transition:
```

Attempting to be more concise

> 1. **Backup Existing Index Databases**

Reviewer 2: Maybe be extra crisp here that the backup should be done while the node is stopped

Reviewer: Good callout.  Maybe just have an explicit step for "Stop your node".

Author: This is done.

>    - Before restarting your Lotus node, create a backup of your existing index databases.
>    - These files are located in the `{$LOTUS_PATH/sqlite}` directory.
>    - While not used in the migration process, these backups are crucial for potential rollbacks.

Reviewer: ```suggestion
   - Before restarting your Lotus node, create a backup of your existing index databases.
   - While not used in the migration process, these backups are crucial for potential rollbacks.
```bash
cd {$LOTUS_PATH/sqlite}
for f in MsgIndex EthTxIndex EventIndex; do
  cp $f $f-backup;
done
```

The idea here is to make it even quicker to follow with copy/pasting

2. **Remove Old Index Files**
   - After creating backups, remove the SQLite database files for `MsgIndex`, `EthTxIndex`, and `EventIndex` from the `{$LOTUS_PATH/sqlite}` directory.

3. **Update Configuration**
   - Modify your Lotus configuration to enable the `ChainIndexer` as described in the `ChainIndexer Config` section above.

4. **Restart Lotus Node**
   - Restart your Lotus node with the new configuration.

>    - The `ChainIndexer` will begin indexing **real-time chain state changes** immediately.

Reviewer 2: What about the migration time when moving to the ChainIndexer? Do we have some estimates here

Reviewer 2: Or to be clearer, it now sound a little bit like:

- Create a backup
- Remove Old Index
- Update Config
- Restart Lotus Node

And you should be all good. But for both snapshot synced / archival node operators it will take quite some time to do the backfilling?

Author: @rjan90 Noted. Will add the timings when I have them.

Reviewer: I'm marking this as a TODO in https://github.com/filecoin-project/lotus/pull/12600.  I'm resolving it her so don't have to keep opening up this big PR to fine unresolved conversations.

> Once Lotus starts with the `ChainIndexer` enabled, it will begin indexing real-time chain state changes i.e. new incoming tipsets. However, it will not index any historical chain state i.e. any previously existing chain state. To index historical chain state (aka **"backfilling"**), you can use the following tools that we're shipping with the `ChainIndexer`:

Reviewer: ```suggestion
### Backfilling
Once Lotus starts with the `ChainIndexer` enabled, it will begin indexing real-time chain state changes (i.e., new incoming tipsets). However, it will not index any historical chain state (i.e., any previously existing chain state). To index historical chain state (i.e., **"backfilling"**), you can use the following tools.
```

I think backfilling can have it's own section under "Migration Guide"

> ### The `ChainValidateIndex` JSON RPC API

Reviewer: ```suggestion
#### The `ChainValidateIndex` JSON RPC API
```

Assuming we made a `### Backfilling` section.

> The `ChainValidateIndex` JSON RPC API serves a dual purpose: it validates/diagnoses the integrity of the index at a specific epoch (i.e., it ensures consistency between indexed data and actual chain state), while also providing the option to backfill the `ChainIndexer` if it does have data for the specified epoch. 

Reviewer: I think you have a better more complete writeup below when you discuss "The `ChainValidateIndex` API serves multiple purposes".  I would scrap this for that one.

```go
// IndexValidation contains detailed information about the validation status of a specific chain epoch.

> type IndexValidation struct {

Reviewer: I assume this is all copy/paste from source code.  I worry it will get out of date.  Can we link to the canonical source that has this information?  Should we link to api-v1-unstable-methods.md and make sure all these comments show up there?

Also, we're talking about this in terms of Go code, but consumers of this are thinking about this in terms of json-rpc right?  Can we talk about it in terms of json-rpc types?

	// TipSetKey is the key of the canonical tipset for this epoch.
	TipSetKey TipSetKey
	// Height is the epoch height at which the validation is performed.
	Height uint64
	// IndexedMessagesCount indicates the number of indexed messages for the canonical tipset at this epoch.
	IndexedMessagesCount uint64
	// IndexedEventsCount signifies the number of indexed events for the canonical tipset at this epoch.
	IndexedEventsCount uint64

> 	// Backfilled denotes whether missing data was successfully backfilled during validation.

Reviewer: ```suggestion
	// Backfilled denotes whether missing data was successfully backfilled into the index during validation.
```

(nit) for clarity?

	Backfilled bool
	// IsNullRound indicates if the epoch corresponds to a null round and therefore does not have any indexed messages or events.
	IsNullRound bool
}

> // ChainValidateIndex validates the integrity of the chain index at a specified epoch and also optionally backfills missing data.

Reviewer: Same comments above of lets link to the canonical source for this info and ideally discuss this in term of json-rpc rather than Go.

//
// Parameters:
//   - epoch: The specific chain epoch for which to validate/backfill the index.
//   - backfill: A boolean flag indicating whether to attempt backfilling of missing data if the index does not have data for the 
//               specified epoch.
//
// Returns:
//   - *types.IndexValidation: A pointer to an IndexValidation struct containing the results of the validation/backfill.
//   - error: An error object if the validation/backfill fails. The error message will contain details about the index 
//            corruption if the call fails because of an incosistency between indexed data and the actual chain state.
//
// Note: The API returns an error if the index does not have data for the specified epoch and backfill is set to false.
func ChainValidateIndex(ctx context.Context, epoch abi.ChainEpoch, backfill bool) (*types.IndexValidation, error)
```

The `ChainValidateIndex` API serves multiple purposes:

1. Validates the chain index at a specific epoch:
   - Ensures consistency between indexed data and actual chain state

>    - Reports any errors found during validation

Reviewer: Newbie question: is this different than "detecting holes" as discussed below?

Author: @BigLep Yes, index corruption could also be:

1) We only have reverted entries for an epoch but no non-reverted entries.
2) We have more than one non-reverted tipset for a given epoch.
3) The number of indexed events for a give tipset does NOT match what's in the chain

and so on and so forth. Basically, "holes" just means missing entries but we could have index corruption other than "holes" too.

2. Optionally backfills missing data:
   - Backfills data if the index is missing information for the specified epoch
   - Backfilling only occurs when the `backfill` parameter is set to `true`

3. Detects "holes" in the index:

>    - If `backfill` is `false` and the index lacks data for the specified epoch, the API returns an error indicating missing data

Reviewer: Maybe inline this into the code itself since it's a good overview comment?

> This API is available for use once the Lotus daemon has started with the `ChainIndexer` enabled. However, calling the API for a single epoch at a time can be cumbersome, especially when backfilling or validating the index over a range of historical epochs, such as during a migration. 
> 
> To simplify this process, we're also providing a command-line tool in `lotus-shed`.
> 
> ### The `lotus-shed chainindex validate-backfill` command-line tool
> 
> **Note: This command can only be run when the Lotus daemon is already running with the `ChainIndexer` enabled as it depends on the `ChainValidateIndex` RPC API described above.**
> 
> The `lotus-shed chainindex validate-backfill` command is a tool for validating and optionally backfilling the chain index over a range of epochs. It wraps the `ChainValidateIndex` API to efficiently process multiple epochs.

Reviewer: ```suggestion
The `ChainValidateIndex` RPC API  is available for use once the Lotus daemon has started with [`ChainIndexer` enabled](#link-to-section). 

#### `lotus-shed chainindex validate-backfill` tool
The `lotus-shed chainindex validate-backfill` command is a tool for validating and optionally backfilling the chain index over a range of epochs since calling the API for a single epoch at a time can be cumbersome, especially when backfilling or validating the index over a range of historical epochs, such as during a migration. It wraps the `ChainValidateIndex` API to efficiently process multiple epochs.

**Note: This command can only be run when the Lotus daemon is already running with the [`ChainIndexer` enabled](#link-to-appropriate-sectoin) as it depends on the [`ChainValidateIndex` RPC API](#link-to-appropriate-section).**
```

Attempting to make this a bit more concise and to be discussing `lotus-shed chainindex validate-backfill` in the `lotus-shed chainindex validate-backfill` section.

> The `lotus-shed chainindex validate-backfill` command is a tool for validating and optionally backfilling the chain index over a range of epochs. It wraps the `ChainValidateIndex` API to efficiently process multiple epochs.

Reviewer: I'm marking this as a TODO in https://github.com/filecoin-project/lotus/pull/12600.  I'm resolving it her so don't have to keep opening up this big PR to fine unresolved conversations.

#### Usage:
```
lotus-shed chainindex validate-backfill --from <start_epoch> --to <end_epoch> [--backfill] [--log-good]
```

#### Parameters:
- `--from` (required): The starting epoch (inclusive) for the validation range. Must be greater than 0.

> - `--to` (required): The ending epoch (inclusive) for the validation range. Must be greater than 0 and less than or equal to `from`.

Reviewer: Newbie here: I assume it's standard for `from` > `to` and that we walk balk from `from` to `to`?

I assume it's not feasible, easy, or sensible to walk from `from` to `to` regardless of which one is greater than the other?  (Don't feel like you need to explain it to me why this is a bad idea - I trust you - I just wanted to ask to make sure.)

Author: @BigLep 

- Great question. Given our implementation of the migration RPC/tooling -> the walk can be in either direction because the RPC/tooling handles forks by explicitly throwing/logging errors asking users to retry. 
- However, the nice thing about walking from `X` to `Y` where `X >Y` is that you will not run into chain forks because if a tipset `Tx` at epoch `X` is on the _"canonical"_ chain(the main chain) -> it follows that `Tipset(Parent(Tx)) i.e. Tx-1` is also on the canonical chain because the hash of a parent is part of the content addressing for a block.
- The same cannot be true when going from epoch `Y` to `X` here it might turn out that the "next" tipset you fetch from the chain state might be the child of a different tipset than the tipset you saw before you asked the chain state for the next tipset.

Reviewer: Thanks - sounds good.  I'll resolve the conversation.

- `--backfill` (optional, default: true): Whether to backfill missing index entries during validation. 
- `--log-good` (optional, default: false): Whether to log details for tipsets that have no detected problems.

> The command validates the chain index entries for each epoch in the specified range, checking for missing or inconsistent entries. If `--backfill` is enabled (which it is by default), it will attempt to backfill any missing entries using the `ChainValidateIndex` API.

Reviewer: This feels like an overview/summary statement that should be listed before the "parameters".

#### Error conditions:
- If `from` or `to` are invalid (<=0 or `to` > `from`), an error is returned.
- If the `ChainValidateIndex` API returns an error for an epoch, indicating an inconsistency between the index and chain state, an error message is logged for that epoch.

#### Logging:

> - **Progress is logged every 2880 epochs (1 day worth of epochs) during the validation process.**

Reviewer: ```suggestion
- **Progress is logged every 2880 epochs (1 day worth of epochs) processed during the validation process.**
```

(nit) mkaing clear that this isn't wall-clock time but rather based on how much data is processed.

> - If `--log-good` is enabled, details are also logged for each epoch that has no detected problems. This includes:

Reviewer: It sounds like it's fair to say that with `--log-good`, one gets one log entry per epoch regardless if is missing data, null, or valid.  Is that right?

Author: @BigLep Not for missing data. That's an error. Let me rewrite this line.

Reviewer: What I'm meaning is that since "--log-good" adds extra logging beyond the normal errors that will be logged, doesn't using "--log-good" cause someone to now see one log entry per epoch?  

> if the epoch is NOT a "Filecoin null round" but we still don't have any data for it in the Index. That's an error.

Without "--log-good" we log in that case right?  And it's still logged even if "--log-good" is used?

I'm probably splitting hairs here.  This is more me trying to make sure I have understanding.  Feel free to resolve.

  - Null rounds with no messages/events.

>   - Epochs with a valid indexed entry.

Reviewer: I think it would be better to have this logged in the help/usage message of the tool itself rather than getting snapshotted here.  In this doc we can direct someone to type `lotus-shed chainindex validate-backfill -h` for informative usage directions.

> #### Example usage:

Reviewer: Example usage seems like a good thing to have in this guide, even if we push canonical information into the command help text text (although you could put it there instead as well).

To validate and backfill the chain index for the last 5760 epochs (2 days) and log details for all epochs:

```
lotus-shed chainindex validate-backfill --from 1000000 --to 994240 --log-good 
```

This command is useful for backfilling the chain index over a range of historical epochs during the migration to the new `ChainIndexer`. **It can also be run periodically to validate the index's integrity.**

> Please free to ask questions on `#fil-lotus-dev` on Filecoin Slack or create issues on Lotus [GitHub](https://github.com/filecoin-project/lotus/issues) for any questions/bugs/comments/concerns.

Reviewer: ```suggestion
## Need more help?
Please free to ask questions on `#fil-lotus-dev` on Filecoin Slack or create issues on Lotus [GitHub](https://github.com/filecoin-project/lotus/issues) for any questions/bugs/comments/concerns.
```

Put this under a section so it shows up in the table of contents.
